examples/core/rgcn/hetero_rgcn.py

```diff
@@ -0,0 +1,654 @@
+                activation=F.relu,
+                dropout=self.dropout,
+            )
+        )
+
+        # Second layer: transform hidden features to output features. No
+        # activation function is applied at this stage.
+        self.layers.append(
+            RelGraphConvLayer(
+                self.hidden_size,
+                self.out_dim,
+                g.ntypes,
+                self.relation_names,
+                activation=None,
+            )
+        )
+
+    def reset_parameters(self):
+        # Reset the parameters of each layer.
+        for layer in self.layers:
+            layer.reset_parameters()
+
+    def forward(self, h, blocks):
+        for layer, block in zip(self.layers, blocks):
+            h = layer(block, h)
+        return h
+
+
+class Logger(object):
+    r"""
+    This class was taken directly from the PyG implementation and can be found
+    here: https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppre
+    d/mag/logger.py
+
+    This was done to ensure that performance was measured in precisely the same
+    way
+    """
+
+    def __init__(self, runs):
+        self.results = [[] for _ in range(runs)]
+
+    def add_result(self, run, result):
+        assert len(result) == 3
+        assert run >= 0 and run < len(self.results)
+        self.results[run].append(result)
+
+    def print_statistics(self, run=None):
+        if run is not None:
+            result = 100 * th.tensor(self.results[run])
+            argmax = result[:, 1].argmax().item()
+            print(f"Run {run + 1:02d}:")
+            print(f"Highest Train: {result[:, 0].max():.2f}")
+            print(f"Highest Valid: {result[:, 1].max():.2f}")
+            print(f"  Final Train: {result[argmax, 0]:.2f}")
+            print(f"   Final Test: {result[argmax, 2]:.2f}")
+        else:
+            result = 100 * th.tensor(self.results)
+
+            best_results = []
+            for r in result:
+                train1 = r[:, 0].max().item()
+                valid = r[:, 1].max().item()
+                train2 = r[r[:, 1].argmax(), 0].item()
+                test = r[r[:, 1].argmax(), 2].item()
+                best_results.append((train1, valid, train2, test))
+
+            best_result = th.tensor(best_results)
+
+            print(f"All runs:")
+            r = best_result[:, 0]
+            print(f"Highest Train: {r.mean():.2f} ± {r.std():.2f}")
+            r = best_result[:, 1]
+            print(f"Highest Valid: {r.mean():.2f} ± {r.std():.2f}")
+            r = best_result[:, 2]
+            print(f"  Final Train: {r.mean():.2f} ± {r.std():.2f}")
+            r = best_result[:, 3]
+            print(f"   Final Test: {r.mean():.2f} ± {r.std():.2f}")
+
+
+def train(
+    g,
+    model,
+    node_embed,
+    optimizer,
+    train_loader,
+    split_idx,
+    labels,
+    logger,
+    device,
+    run,
+):
+    print("start training...")
+    category = "paper"
+
+    for epoch in range(3):
```

Comment:
Why the number of epochs is so small?

Reply:
According to the [README.md](https://github.com/dmlc/dgl/blob/master/examples/pytorch/ogb/ogbn-mag/README.md).

> Typically, the best Validation performance is obtained after the 1st or 2nd epoch, after which it begins to overfit. This is why "Highest Train" (typically occuring at the end of the 3rd epoch), is significantly higher than "Final Train" (corresponding to epoch of maximal Validation performance).